itou/www/apply/views/submit_views.py

```diff
@@ -86,63 +68,120 @@ def get_approvals_wrapper(request, job_seeker, siae):
-        raise PermissionDenied("Vous ne pouvez postuler pour un candidat que dans votre structure.")
-
-    # Refuse all applications except those issued by the SIAE
-    if siae.block_job_applications and not siae.has_member(request.user):
-        # Message only visible in DEBUG
-        raise Http404("Cette organisation n'accepte plus de candidatures pour le moment.")
-
-    back_url = get_safe_url(request, "back_url")
-
-    # Start a fresh session.
-    request.session[settings.ITOU_SESSION_JOB_APPLICATION_KEY] = {
-        "back_url": back_url,
-        "job_seeker_pk": None,
-        "nir": None,
-        "siae_pk": siae.pk,
-        "sender_pk": None,
-        "sender_kind": None,
-        "sender_siae_pk": None,
-        "sender_prescriber_organization_pk": None,
-        "job_description_id": request.GET.get("job_description_id"),
-    }
+        return decorated
 
-    next_url = reverse("apply:step_sender", kwargs={"siae_pk": siae.pk})
-    return HttpResponseRedirect(next_url)
+    return wrapper
 
 
-@login_required
-@siae_session_required
-def step_sender(request, siae_pk):
-    """
-    Determine info about the sender.
-    """
-    user_info = get_user_info(request)
+siae_session_required = valid_session_required(["siae_pk"])
 
-    session_data = request.session[settings.ITOU_SESSION_JOB_APPLICATION_KEY]
-    session_data["sender_pk"] = user_info.user.pk
-    session_data["sender_kind"] = user_info.kind
 
-    if user_info.prescriber_organization:
-        session_data["sender_prescriber_organization_pk"] = user_info.prescriber_organization.pk
+class SessionDataBaseMixin(LoginRequiredMixin, ContextMixin):
+    """A mixin that helps the programmer to handle its session data, through a dedicated session key.
+    It also takes care of re-injecting a "back_url" to the context if it was saved in it.
+    """
+
+    def get_session_key(self):
+        raise NotImplementedError()
+
+    def get_session(self, key=None):
+        session_data = self.request.session[self.get_session_key()]
+        if key:
+            return session_data[key]
+        return session_data
+
+    def init_session(self, data):
+        self.request.session[self.get_session_key()] = data
+
+    def set_session(self, key, value):
+        session_data = self.get_session()
+        session_data[key] = value
+        self.request.session.modified = True
+
+    def get_back_url(self):
+        session_back_url = self.get_session().get("back_url")
+        if session_back_url:
+            return get_safe_url(request=self.request, url=session_back_url)
+        return None
+
+    def get_context_data(self, **kwargs):
+        return super().get_context_data(**kwargs) | {
+            "back_url": self.get_back_url(),
+        }
+
+    def dispatch(self, request, *args, **kwargs):
+        if not self.get_session():
+            # please take care of initializing your session data with init_session() in
```

Comment:
Please take care of adding a capital letter at the beginning of this sentence. My brain and my eyes will thank you! :)